// appium-windows-driver/build/lib/commands/log.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedLogTypes = void 0;
const lodash_1 = __importDefault(require("lodash"));
const COLOR_CODE_PATTERN = /\u001b\[(\d+(;\d+)*)?m/g; // eslint-disable-line no-control-regex
const GET_SERVER_LOGS_FEATURE = 'get_server_logs';
const DEFAULT_LOG_LEVEL = 'ALL';
exports.supportedLogTypes = {
    server: {
        description: 'Appium server logs',
        getter: (self) => {
            self.assertFeatureEnabled(GET_SERVER_LOGS_FEATURE);
            return self.log.unwrap().record.map(nativeLogEntryToSeleniumEntry);
        },
    },
};
function nativeLogEntryToSeleniumEntry(x) {
    const msg = lodash_1.default.isEmpty(x.prefix) ? x.message : `[${x.prefix}] ${x.message}`;
    return toLogEntry(lodash_1.default.replace(msg, COLOR_CODE_PATTERN, ''), x.timestamp ?? Date.now());
}
function toLogEntry(message, timestamp, level = DEFAULT_LOG_LEVEL) {
    return { timestamp, level, message };
}
